sdk/core/azure-core/src/http/url.cpp

```diff
@@ -103,86 +103,21 @@ std::string Url::Decode(const std::string& value)
   return decodedValue;
 }
 
-static const char* unreserved
-    = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~";
-static const char* subdelimiters = "!$&'()*+,;=";
-
-std::string Url::EncodeHost(const std::string& host)
-{
-  return EncodeImpl(host, [](int c) { return c > 127; });
-}
-
-std::string Url::EncodePath(const std::string& path)
-{
-  const static std::vector<bool> shouldEncodeTable = []() {
-    const std::string pathCharacters
-        = std::string(unreserved) + std::string(subdelimiters) + "%/:@";
-
-    std::vector<bool> ret(256, true);
-    for (char c : pathCharacters)
-    {
-      ret[c] = false;
-    }
-    // we also encode % and +
-    ret['%'] = true;
-    ret['+'] = true;
-    return ret;
-  }();
-
-  return EncodeImpl(path, [](int c) { return shouldEncodeTable[c]; });
-}
-
-std::string Url::EncodeQuery(const std::string& query)
-{
-  const static std::vector<bool> shouldEncodeTable = []() {
-    std::string queryCharacters = std::string(unreserved) + std::string(subdelimiters) + "%/:@?";
-
-    std::vector<bool> ret(256, true);
-    for (char c : queryCharacters)
-    {
-      ret[c] = false;
-    }
-    // we also encode % and +
-    ret['%'] = true;
-    ret['+'] = true;
-    // Surprisingly, '=' also needs to be encoded because Azure Storage server side is so strict.
-    // We are applying this function to query key and value respectively, so this won't affect
-    // that = used to separate key and query.
-    ret['='] = true;
-    return ret;
-  }();
-
-  return EncodeImpl(query, [](int c) { return shouldEncodeTable[c]; });
-}
-
-std::string Url::EncodeFragment(const std::string& fragment)
-{
-  const static std::vector<bool> shouldEncodeTable = []() {
-    std::string queryCharacters = std::string(unreserved) + std::string(subdelimiters) + "%/:@?";
-
-    std::vector<bool> ret(256, true);
-    for (char c : queryCharacters)
-    {
-      ret[c] = false;
-    }
-    // we also encode % and +
-    ret['%'] = true;
-    ret['+'] = true;
-    return ret;
-  }();
-
-  return EncodeImpl(fragment, [](int c) { return shouldEncodeTable[c]; });
-}
-
-std::string Url::EncodeImpl(const std::string& source, const std::function<bool(int)>& shouldEncode)
+std::string Url::Encode(const std::string& value, const char* extraSymbolsToEncode)
 {
   const char* hex = "0123456789ABCDEF";
+  std::string extraSymbolsString(extraSymbolsToEncode);
+  std::unordered_set<unsigned char> extraChars(
```

Comment:
Why did you abandon the static vector? This is even slower. You have to construct an `unordered_map` for every single encode operation. Lookup in hash map is expected O(1) and cache-unfriendly, while lookup in vector is strict O(1) and cache-friendly.

Reply:
What's the benefit of this change?

Reply:
Would a static vector in multi-threaded cases when the user provides different `extraSymbolsToEncode` input parameters to this function?

Reply:
why `extraSymbolsToEncode` would be different?

Reply:
> why `extraSymbolsToEncode` would be different?

Thread 1, calls `Encode("foo=", "=")`, expected result should be: "foo="
Thread 2, calls `Encode("bar=")`, expected result should be: "bar%3D"

What would a static `extraChars` contain in that case? There is no locking/synchronization mechanism. Will we get deterministic/expected behavior if we used a static vector?

Reply:
OK, I see. We used to have different `Encode` functions for path, query parameter, fragment, etc. Each functions had its own encode characters table. Victor has changed all those into one function. So we now we only have one table.

I really cannot see any benefits of doing so, except making encoding slower. But whatever, as long as it works.

Reply:
The `defaultNonUrlEncodeChars` set is static and contains all the reserved chars that we don't want to encode.

Only if user pass some argument for `doNotEncodeSymbols`, then a string is created and then a set for it.
It might not be a huge difference since the length of the customer input is not expected to be that big